mathics/eval/arithmetic.py

```diff
@@ -0,0 +1,812 @@
+# -*- coding: utf-8 -*-
+
+"""
+helper functions for arithmetic evaluation, which do not
+depends on the evaluation context. Conversions to Sympy are
+used just as a last resource.
+"""
+
+from typing import List, Optional, Tuple
+
+import mpmath
+import sympy
+
+from mathics.core.atoms import (
+    Complex,
+    Integer,
+    Integer0,
+    Integer1,
+    IntegerM1,
+    Number,
+    Rational,
+    RationalOneHalf,
+    Real,
+)
+from mathics.core.convert.mpmath import from_mpmath
+from mathics.core.convert.sympy import from_sympy
+from mathics.core.element import BaseElement
+from mathics.core.expression import Expression
+from mathics.core.number import dps, min_prec
+from mathics.core.symbols import Atom
+from mathics.core.systemsymbols import (
+    SymbolAbs,
+    SymbolE,
+    SymbolEulerGamma,
+    SymbolI,
+    SymbolKhinchin,
+    SymbolLog,
+    SymbolPi,
+    SymbolPlus,
+    SymbolPower,
+    SymbolTimes,
+)
+from mathics.eval.numbers import eval_accuracy
+
+RationalMOneHalf = Rational(-1, 2)
+RealM0p5 = Real(-0.5)
+RealOne = Real(1.0)
+
+
+# Here we store numerical constants with their approximate
+# numerical values.
+
+NUMERICAL_CONSTANTS = {
+    SymbolE: 2.718281828,
+    SymbolEulerGamma: 0.5772156649,
+    SymbolKhinchin: 2.685452001,
+    SymbolPi: 3.141592654,
+}
+
+
+# Here I used the convention of calling eval_* to functions that can produce a new expression, or None
```

Comment:
@rocky, I leave this long comment here about the eval<-> apply semantics. Maybe we want to go over this again before merging this.

Reply:
I looked at this and there is information here that I a do not believe is elsewhere. I think it should be elsewhere but I don't know right now _where_ that elsewhere is.